### ISSUE TITLE
Allow disabling/comenting CSV or Google Sheet row

### ISSUE THREAD

**Issue by Digital-Grinnell** [mjordan/islandora_workbench#203]
Another trick I introduced in my fork of IMI...

I find lots of occasions when I want to put a comment line into a CSV or Google Sheet just for record-keeping purposes, or simply to preserve a line of data without fear of ingesting it by mistake.  We do that in Digital.Grinnell's copy of IMI by stipulating that any row beginning with a # (hashtag) is treated as a comment and skipped during ingest.

The following line is from the "Instructions" tab of our MASTER IMI Google Sheet template...

> Place a hashtag (#) in the first character of the first cell (column A) in a row to "disable" import of that entire row/record.  Common practice is to label the first column 'PID' and record the PID of an already ingested record within, something like "# test:12345". 

I understand that column-order is not significant in Islandora_Workbench, but neither is it in IMI.  This hashtag capability is implemented in IMI as rows are read in from the input where the entire row is captured as a string.  There's a simple test in the code to check if the first character in the string is a hashtag, if it is, the loop that reads CSV rows simply skips forward.  This is not ideal, because the comment line is effectively "lost" during processing and subsequently does not appear in any output or logs the process might generate.

**Comment by mjordan** [comment]
Workbench uses Python's standard CSV parser library, which doesn't treat CSV data as rows of strings, but the test you describe could easily be applied to the value of the "id" (or equivalent) field, e.g. 

```
id,title
#0002,My title
0003,My other title
```

or, if "id" is not the first column in the CSV,

```
title,id
My title,#0002
My other title,0003
```

Would that work?

**Comment by mjordan** [comment]
`#` at the start of each CSV will may work, since the CSV reader treats each row as an "ordered dictionary", so the first column is always the first entry in the dictionary. In other words, the `#` wouldn't need to apply to the `id` field as I describe above. This makes more sense.

**Comment by mjordan** [comment]
After a bunch of googling, it's clear that the standard way of filtering out rows using Pythons csv library is to read the file in, filter the rows, and write the file out again, since csv Readers and Writers need to take a file handle as their first argument. I'll continue to poke around to see if it's possible to avoid writing out the modified file by using `io.StringIO()` instead of an on-disk file, but Workbench's `get_csv_data()` already writes out a modified CSV file if `csv_field_templates` is present in the config. Before doing many more acrobatics with `io.StringIO()` I'd like to see if I build on the existing code.

**Comment by mjordan** [comment]
@Digital-Grinnell you can now comment out rows in your input CSV, Excel file, or Google Sheet by placing a `#` at the start of the first column. Here are the accompanying docs: https://mjordan.github.io/islandora_workbench_docs/fields/#commenting-out-csv-rows .

Commented rows are skipped entirely, so they don't appear in logs, etc. I'm not exactly sure what you mean by "the comment line is effectively 'lost' during processing and subsequently does not appear in any output or logs the process might generate". If you'd like to elaborate, please do so!

**Comment by Digital-Grinnell** [comment]
Sounds good, @mjordan.  Occasionally I'll put meaningful, instructive comments in such a row so it would be nice if they were echoed/preserved in `google_sheet.csv` and elsewhere, but I understand that can be difficult to achieve.  Thanks for implementing this as you have, I think it will be a welcome improvement.  Take care.

**Comment by mjordan** [comment]
Thanks for the additional info. It might be possible to retain them in the local CSV dump. I'll take a look during the next couple of days.

**Comment by mjordan** [comment]
@Digital-Grinnell commented rows in the Google Sheet remain in the "google_sheet.csv" file, but the way I've implemented removing the comments, and adding CSV field templates, is to generate a CSV file names after the input_csv with the extension ".preprocessed". This is the file that Workbench actually uses to populate Drupal.

TLDR - if you comment out a row in your Google Sheet, or add an instructive comment as a row, it will show up in "google_sheet.csv."

**Comment by mjordan** [comment]
@Digital-Grinnell OK to close this issue?

**Comment by Digital-Grinnell** [comment]
Yes, thanks Mark.

Sent from my iPad

On Feb 28, 2021, at 6:36 PM, Mark Jordan <notifications@github.com> wrote:

﻿

@Digital-Grinnell<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Digital-2DGrinnell&d=DwMCaQ&c=HUrdOLg_tCr0UMeDjWLBOM9lLDRpsndbROGxEKQRFzk&r=D8E-oGNaPT9srWV6jE8UP5unsmKEmmHEH-tzgmjBvLk&m=iUiJsT9vZnFUsTMBvZdnhFwKNRIrQyIv6W-y9jUvUT0&s=FEeSsw24dHrWx1E8zvCMeoy9HMVDSbPzzvxX73afShU&e=> OK to close this issue?
